internal/controller/sync.go

```diff
@@ -537,128 +458,127 @@ func (r *Route) buildNextCR(ctx context.Context, route *routev1.Route, revision
 			return nil, err
 		}
 	}
+
 	var serialNumber string
 	if metav1.HasAnnotation(route.ObjectMeta, cmapi.SubjectSerialNumberAnnotationKey) {
 		serialNumber = route.Annotations[cmapi.SubjectSerialNumberAnnotationKey]
 	}
-	csr, err := x509.CreateCertificateRequest(
-		rand.Reader,
-		&x509.CertificateRequest{
-			Version:            0,
-			SignatureAlgorithm: signatureAlgorithm,
-			PublicKeyAlgorithm: publicKeyAlgorithm,
-			Subject: pkix.Name{
-				CommonName:         route.Annotations[cmapi.CommonNameAnnotationKey],
-				Country:            countries,
-				Locality:           localities,
-				Organization:       organizations,
-				OrganizationalUnit: organizationalUnits,
-				PostalCode:         postalCodes,
-				Province:           provinces,
-				SerialNumber:       serialNumber,
-				StreetAddress:      streetAddresses,
-			},
-			EmailAddresses: emailAddresses,
-			DNSNames:       dnsNames,
-			IPAddresses:    ipSans,
-			URIs:           uriSans,
-		},
-		key,
-	)
-	if err != nil {
-		return nil, err
-	}
-	csrPEM := pem.EncodeToMemory(&pem.Block{
-		Type:  "CERTIFICATE REQUEST",
-		Bytes: csr,
-	})
 
-	var issuerName string
-	if metav1.HasAnnotation(route.ObjectMeta, cmapi.IngressIssuerNameAnnotationKey) {
-		issuerName = route.Annotations[cmapi.IngressIssuerNameAnnotationKey]
-	} else {
-		issuerName = route.Annotations[cmapi.IssuerNameAnnotationKey]
+	var revisionHistoryLimit *int32
+	if metav1.HasAnnotation(route.ObjectMeta, cmapi.RevisionHistoryLimitAnnotationKey) {
+		historyLimitRaw := route.Annotations[cmapi.RevisionHistoryLimitAnnotationKey]
+
+		parsedLimit, err := strconv.ParseInt(historyLimitRaw, 10, 32)
+		if err != nil {
+			return nil, fmt.Errorf("invalid revision-history-limit annotation %q on %s/%s", historyLimitRaw, route.Namespace, route.Name)
+		}
+
+		typedLimit := int32(parsedLimit)
+		revisionHistoryLimit = &typedLimit
 	}
 
-	cr := &cmapi.CertificateRequest{
+	secretName := route.Name + "-tls-cert"
```

Comment:
There might be a risk of hitting the max name length limit here.

Reply:
Yeah I've confirmed this by changing the smoke test to create a route with a longer name:

```patch
index bd3ed6c..0c9423e 100755
--- a/test/test-smoke.sh
+++ b/test/test-smoke.sh
@@ -58,7 +58,7 @@ EOF
 
 # Create a Route and patch the status with multiple hosts
 
-route_name="test-"$(openssl rand -hex 12)
+route_name="test-"$(openssl rand -hex 124)
 
 cat <<EOF | kubectl create -f -
 apiVersion: route.openshift.io/v1
```

I was going to just raise an issue for this but I actually don't want to merge with this in; I'll change.

Reply:
This is now fixed along with plenty of tests!